### PR TITLE
#249: Non-Intrusive examples and docs

### LINKED ISSUE #249: Improve non-intrusive serialization documentation and examples
Current examples are intrusive. Add non-intrusive examples. Be sure to make it clear that the non-intrusive serializer must go in ~~the checkpoint (or later magistrate) namespace~~ the same namespace as the type being serialized.

### PR DESCRIPTION
**Pull request by thearusable**
Closes #249

### CODE REVIEW

File: examples/checkpoint_example_traversal.cc [hunk 1]
```diff
@@ -149,3 +151,5 @@ int main(int, char**) {
 
   return 0;
 }
+
+/// [Serialization with custom traverser]
```
Comment (PhilMiller) [inline examples/checkpoint_example_traversal.cc]: Please fix the end-of-file newline, so utilities don't complan

Comment (thearusable) [reply examples/checkpoint_example_traversal.cc]: Ok Fixed.

File: docs/md/checkpoint_learn_example_2_nonintrusive.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+\page ckpt_learn_ex2_nonintrusive Non-Intrusive Program Example 2
+
+\attention All Non-Intrusive serialize methods <b>MUST</b> be placed in the namespace of type which they serialize.
+
+The full code for this *checkpoint* example can be found here:
```
Comment (cz4rs) [inline docs/md/checkpoint_learn_example_2_nonintrusive.md:5]: I guess that's `magistrate` now.
```suggestion
The full code for this *magistrate* example can be found here:
```

Comment (cz4rs) [reply docs/md/checkpoint_learn_example_2_nonintrusive.md:5]: Although, on the other hand, there's still `checkpoint` everywhere in the file names and the code :shrug:

Comment (cz4rs) [reply docs/md/checkpoint_learn_example_2_nonintrusive.md:5]: Let's stick with `checkpoint` for consistency with the other examples.

File: examples/checkpoint_example_polymorphic_macro_nonintrusive.cc [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    assert(val_ == 30);
+    assert(a==10);
+    assert(b==20);
+    assert(c==100);
+  }
+};
+
+/*
+ * Example vector that holds a vector of unique_ptr to MyBase
+ */
```
Comment (cz4rs) [inline examples/checkpoint_example_polymorphic_macro_nonintrusive.cc]: The comment here seems to be redundant, this is pretty self-explanatory.

File: examples/checkpoint_example_to_file_nonintrusive.cc [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+static constexpr int const u_val = 934;
+
+// \struct MyTestType
+// \brief Simple structure with two variables of built-in types
+struct MyTestType {
+
+  // \brief Default constructor
+  //
+  // The default constructor is needed for the (de)serialization.
+  // (required for serialization)
```
Comment (cz4rs) [inline examples/checkpoint_example_to_file_nonintrusive.cc]: These 2 lines state pretty much the same thing.

File: examples/checkpoint_example_to_file_nonintrusive.cc [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  //
+  // De-serializes from the file an object of type 'MyTestType'
+  // out will be an object of type 'std::unique_ptr<MyTestType>'
+  //
+  auto out = checkpoint::deserializeFromFile<MyTestType>("hello.txt");
+
+  if (my_test_inst == *out)
+    std::cout << " Serialization / Deserialization from file worked. \n";
+  else
+    std::cout << " Serialization / Deserialization from file failed. \n";
```
Comment (cz4rs) [inline examples/checkpoint_example_to_file_nonintrusive.cc:154]: We could actually use an assertion here to ensure that potential failure is visible in CI.

File: examples/checkpoint_example_2_nonintrusive.cc [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  MyTest2 my_test_2;
+
+  // \brief Default constructor
+  //
+  // The reconstruction strategy is required for deserialization. A default
+  // constructor is one of the reconstruction strategies that checkpoint will
+  // look for.
+  MyTest() = default;
+
+  // \brief Printing function unto the standard display
```
Comment (cz4rs) [inline examples/checkpoint_example_2_nonintrusive.cc:86]: Self-explanatory.

**Review by cz4rs — COMMENTED** [review]
Looks pretty good overall, I'm going to give it another look after the weekend.
Some nitpicking remarks about the comments, but I don't feel like these *have to* be addressed.

**Review by cz4rs — APPROVED** [review]

File: examples/checkpoint_example_polymorphic_macro_nonintrusive.cc (hunk 3 above)
Comment (thearusable) [reply examples/checkpoint_example_polymorphic_macro_nonintrusive.cc]: Ok, I removed it

File: examples/checkpoint_example_to_file_nonintrusive.cc (hunk 4 above)
Comment (thearusable) [reply examples/checkpoint_example_to_file_nonintrusive.cc]: Indeed, I removed the `(required for serialization)`

File: examples/checkpoint_example_to_file_nonintrusive.cc (hunk 5 above)
Comment (thearusable) [reply examples/checkpoint_example_to_file_nonintrusive.cc:154]: Sure, added

File: examples/checkpoint_example_2_nonintrusive.cc (hunk 6 above)
Comment (thearusable) [reply examples/checkpoint_example_2_nonintrusive.cc:86]: Removed

**Review by cz4rs — APPROVED** [review]

File: examples/checkpoint_example_3_nonintrusive.cc [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  }
+};
+
+}}} // end namespace magistrate::nonintrusive::examples
+
+// \brief In Non-Intrusive way, serialize function needs to be placed in the namespace
+// of the type which will be serialized.
+namespace magistrate { namespace nonintrusive { namespace examples {
+
+// \breif Non-Intrusive Serialize method for TestDefaultCons structure.
```
Comment (PhilMiller) [inline examples/checkpoint_example_3_nonintrusive.cc]:
```suggestion
// \brief Non-Intrusive Serialize method for TestDefaultCons structure.
```

File: examples/checkpoint_example_3_nonintrusive.cc [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+//
+// \note Together with default constructor provides a serialization / deserialization
+// capability to the structure.
+template <typename Serializer>
+void serialize(Serializer& s, TestDefaultCons& tdc) {
+  s | tdc.a;
+}
+
+// \breif Non-Intrusive Serialize method for TestShouldFailReconstruct structure.
```
Comment (PhilMiller) [inline examples/checkpoint_example_3_nonintrusive.cc]:
```suggestion
// \brief Non-Intrusive Serialize method for TestShouldFailReconstruct structure.
```

File: examples/checkpoint_example_3_nonintrusive.cc [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+template <typename Serializer>
+void serialize(Serializer& s, TestShouldFailReconstruct& tsf) {
+  s | tsf.a;
+}
+
+// \breif Non-Intrusive Serialize method for TestReconstruct structure.
```
Comment (PhilMiller) [inline examples/checkpoint_example_3_nonintrusive.cc]:
```suggestion
// \brief Non-Intrusive Serialize method for TestReconstruct structure.
```

**Review by PhilMiller — APPROVED** [review]

**Pull request merged**